src/Mod/Sketcher/Gui/DrawSketchHandlerBSpline.h

```diff
@@ -23,618 +23,1151 @@
-                Gui::cmdAppObjectArgs(
-                    sketchgui->getObject(),
-                    "addGeometry(Part.Circle(App.Vector(%f,%f,0),App.Vector(0,0,1),10),True)",
-                    BSplinePoles.back().x,
-                    BSplinePoles.back().y);
+                // Constraint pole/knot circles to B-spline.
+                std::stringstream cstream;
+                cstream << "conList = []\n";
+
+                for (size_t i = 0; i < geoIds.size(); i++) {
+                    cstream << "conList.append(Sketcher.Constraint('InternalAlignment:Sketcher::"
+                               "BSplineControlPoint',"
+                            << geoIds[0] + i << "," << static_cast<int>(Sketcher::PointPos::mid)
+                            << "," << currentgeoid << "," << i << "))\n";
+                }
+
+                cstream << Gui::Command::getObjectCmd(sketchgui->getObject())
+                        << ".addConstraint(conList)\n"
+                        << "del conList\n";
 
-                poleGeoIds.push_back(getHighestCurveIndex());
+                Gui::Command::doCommand(Gui::Command::Doc, cstream.str().c_str());
 
+                // for showing the knots on creation
                 Gui::cmdAppObjectArgs(sketchgui->getObject(),
-                                      "addConstraint(Sketcher.Constraint('Weight',%d,%f)) ",
-                                      poleGeoIds.back(),
-                                      1.0);  // First pole defaults to 1.0 weight
+                                      "exposeInternalGeometry(%d)",
+                                      currentgeoid);
             }
-            catch (const Base::Exception&) {
-                Gui::NotifyError(sketchgui,
-                                 QT_TRANSLATE_NOOP("Notifications", "Error"),
-                                 QT_TRANSLATE_NOOP("Notifications", "Error adding B-Spline pole"));
+            else {
+                unsigned int myDegree = 3;
 
-                Gui::Command::abortCommand();
+                if (!periodic) {
+                    // FIXME: This is hardcoded until degree can be changed
+                    multiplicities.front() = myDegree + 1;
+                    multiplicities.back() = myDegree + 1;
+                }
 
-                static_cast<Sketcher::SketchObject*>(sketchgui->getObject())->solve();
+                std::vector<std::stringstream> streams;
+
+                // Create subsets of points between C0 knots.
+                // The first point
+                streams.emplace_back();
+                streams.back() << "App.Vector(" << points.front().x << "," << points.front().y
+                               << "),";
+                // Middle points
+                for (size_t i = 1; i < points.size() - 1; ++i) {
+                    streams.back() << "App.Vector(" << points[i].x << "," << points[i].y << "),";
+                    if (multiplicities[i] >= myDegree) {

```

Comment:
Can you please resolve these compiler warnings by casting one or the other of these to an appropriate type? (Applies everywhere you do this comparison)

Reply:
Sure I'm on it.